### PR TITLE
Updated regex derivative engine

### PR DESCRIPTION
**Pull request by veanes**
Rewrote/updated several derivate related algos both for correctness and performance.
Main changes involve more efficient use of algebraic laws and avoidance of unnecessary rewrites that in many cases were causing orders of magnitude of overhead. 
Several open issues should be addressed by the updated engine.
Some old code is still there for convenience to debug possible regressions, that will eventually be removed, once any new regression bugs have been flushed out.

### CODE REVIEW

File: src/ast/seq_decl_plugin.cpp [hunk 1]
```diff
@@ -1065,39 +1105,70 @@ app* seq_util::rex::mk_epsilon(sort* seq_sort) {
 /*
   Produces compact view of concrete concatenations such as (abcd).
 */
-std::ostream& seq_util::rex::pp::compact_helper_seq(std::ostream& out, expr* s) const {
+void seq_util::rex::pp::print_seq(std::ostream& out, expr* s) const {
```
Comment (NikolajBjorner) [inline src/ast/seq_decl_plugin.cpp:1108]: returning ostream& allows to chain

File: src/ast/rewriter/seq_rewriter.cpp [hunk 2]
```diff
@@ -908,12 +913,18 @@ expr_ref seq_rewriter::mk_seq_rest(expr* t) {
...
-        result = str().mk_nth_i(s, m_autil.mk_sub(m_autil.mk_add(j, k), one));
+    expr* s, * j, * k, * s_, * l;
+    rational jv;
+    if (str().is_extract(t, s, j, k) && m_autil.is_numeral(j, jv) && jv.is_nonneg()) {
+        unsigned i;
+        // k = |s_| - i, l = |s_| implies that lastpos = j + k - 1 = k + j - 1 = j + l - i - 1 = l + j - i - 1
+        expr_ref lastpos(str().is_len_sub(k, l, s_, i) && s == s_ ? 
+            m_autil.mk_add(l, m_autil.mk_int(jv.get_int32() - 1 - i)) :
+            m_autil.mk_add(k, m_autil.mk_int(jv.get_int32() - 1)), m());
+        result = str().mk_nth_i(s, lastpos);
```
Comment (NikolajBjorner) [inline src/ast/rewriter/seq_rewriter.cpp:924]: use mk_nth_i(s, lastpost.get()) to resolve overload error

File: src/ast/rewriter/seq_rewriter.cpp [hunk 3]
```diff
@@ -2976,29 +2992,377 @@ bool seq_rewriter::check_deriv_normal_form(expr* r, int level) {
...
+        else
+            result = nothing();
+    }
+    else if (re().is_union(r, r1, r2))
+        result = mk_antimirov_deriv_union(mk_antimirov_deriv(e, r1, path), mk_antimirov_deriv(e, r2, path));
+    else if (re().is_intersection(r, r1, r2))
+        result = mk_antimirov_deriv_intersection(
+            mk_antimirov_deriv(e, r1, path),
+            mk_antimirov_deriv(e, r2, path), m().mk_true());
+    else if (re().is_star(r, r1) || re().is_plus(r, r1) || re().is_loop(r, r1, lo) && 0 <= lo && lo <= 1)
```
Comment (NikolajBjorner) [inline src/ast/rewriter/seq_rewriter.cpp:3143]: read build warning message: add parentheses around && sub-expression even though we know && binds stronger.

File: src/ast/rewriter/seq_rewriter.cpp [hunk 4]
```diff
@@ -2976,29 +2992,377 @@ bool seq_rewriter::check_deriv_normal_form(expr* r, int level) {
...
+        result = r;
+    else if (re().is_full_seq(r) && re().is_full_seq(s))
+        result = r;
+    else if (re().is_concat(r, r1, r2))
+        //create the resulting concatenation in right-associative form
+        result = mk_regex_concat(r1, mk_regex_concat(r2, s));
+    else {
+        //TODO: perhaps simplifiy some further cases such as .*. = ..* = .*.+ = .+.* = .+
+        result = re().mk_concat(r, s);
+    }
```
Comment (NikolajBjorner) [inline src/ast/rewriter/seq_rewriter.cpp:3321]: See build warning you need to "return result;"

Comment (veanes) [reply src/ast/rewriter/seq_rewriter.cpp:3321]: fixed this bug and above comments.

File: src/smt/seq_regex.cpp [hunk 5]
```diff
@@ -616,55 +618,57 @@ namespace smt {
...
                 }
-                /*
-                else if (re().is_empty(e))
+                else if (re().is_empty(e) || (s_is_longer_than_i && re().is_epsilon(e)))
                 {
-                    re_to_bool.find(e) = m.mk_false();
+                    // s[i..] in [] <==> false, also: s[i..] in () <==> false when |s|>i
+                    re_to_accept.find(e) = m.mk_false();
                 }
+                else if (re().is_full_seq(e) || s_is_longer_than_i && re().is_dot_plus(e))
```
Comment (NikolajBjorner) [inline src/smt/seq_regex.cpp:639]: [528/845] Building CXX object src/smt/CMakeFiles/smt.dir/seq_regex.cpp.o
/home/vsts/work/1/s/src/smt/seq_regex.cpp:639:68: warning: '&&' within '||' [-Wlogical-op-parentheses]
                else if (re().is_full_seq(e) || s_is_longer_than_i && re().is_dot_plus(e))
                                             ~~ ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/home/vsts/work/1/s/src/smt/seq_regex.cpp:639:68: note: place parentheses around the '&&' expression to silence this warning
                else if (re().is_full_seq(e) || s_is_longer_than_i && re().is_dot_plus(e))
                                                                   ^

Comment (veanes) [reply src/smt/seq_regex.cpp:639]: fixed, will hold off committing yet, until more fixes done, (actually sometimes I do write the parenthesis, just need to be consistent with that) .

**Comment by NikolajBjorner** [comment]
We lost some rewrites

From re.smt2: These should have been rewritten to true:
```
(str.in_re "" (re.* (str.to_re "a")))
(str.in_re "" (re.* (str.to_re "a")))
(str.in_re "" (str.to_re ""))
(not (str.in_re "" (re.+ re.allchar)))
(not (str.in_re "" (re.+ re.allchar)))
````

**Comment by NikolajBjorner** [comment]
re_rewrites.smt2 exposes a reference count leak.

E.g., fix by taking a reference count in callee.

```
 expr_ref seq_rewriter::mk_antimirov_deriv(expr* e, expr* r, expr* path) {
+    expr_ref _path(path, m()), _r(r, ());
     expr_ref result(m_op_cache.find(OP_RE_DERIVATIVE, e, r, path), m());
     if (!result) {
         mk_antimirov_deriv_rec(e, r, path, result);
@@ -3204,6 +3205,7 @@ expr_ref seq_rewriter::mk_antimirov_deriv_intersection(expr* d1, expr* d2, expr*
 
 expr_ref seq_rewriter::mk_antimirov_deriv_concat(expr* d, expr* r) {
     expr_ref result(m());
+    expr_ref _r(r, m()), _d(d, m());
     expr* c, * t, * e;
     if (m().is_ite(d, c, t, e))
         result = m().mk_ite(c, mk_antimirov_deriv_concat(t, r), mk_antimirov_deriv_concat(e, r));
```

File: src/ast/rewriter/seq_rewriter.cpp [hunk 6]
```diff
@@ -3775,8 +4148,10 @@ br_status seq_rewriter::mk_str_in_regexp(expr* a, expr* b, expr_ref& result) {
 
     expr_ref hd(m()), tl(m());
     if (get_head_tail(a, hd, tl)) {
-        result = re().mk_in_re(tl, re().mk_derivative(hd, b));
-        return BR_REWRITE2;
+        //result = re().mk_in_re(tl, re().mk_derivative(hd, b));
+        //result = re().mk_in_re(tl, mk_derivative(hd, b));
+        result = mk_in_antimirov(tl, mk_antimirov_deriv(hd, b, m().mk_true()));
+        return BR_DONE;
```
Comment (NikolajBjorner) [inline src/ast/rewriter/seq_rewriter.cpp:4154]: BR_DONE is incorrect. There could be more rewrites. z3test/regressions/smt2/re.smt2 regresses. It is fixed by returning BR_REWRITE1

Comment (veanes) [reply src/ast/rewriter/seq_rewriter.cpp:4154]: fixed

**Comment by NikolajBjorner** [comment]
bdd "structure" shouldn't matter (much) for of-pred.

File: src/ast/seq_decl_plugin.cpp [hunk 7]
```diff
@@ -1151,100 +1223,209 @@ std::ostream& seq_util::rex::pp::seq_unit(std::ostream& out, expr* s) const {
...
+    }
+    else if (re.m.is_not(e, r1)) {
+        out << "!";
+        print(out, r1);
+    }
+    else if (re.m.is_value(e))
+        out << mk_pp(e, re.m);
+    else if (is_app(e)) {
+        out << "(" << to_app(e)->get_decl()->get_name().str();
+        for (unsigned i = 0; i < to_app(e)->get_num_args(); i++) {
```
Comment (NikolajBjorner) [inline src/ast/seq_decl_plugin.cpp:1397]: for (expr* arg : *to_app(e)) 
    print(out << " ", arg);

File: src/ast/seq_decl_plugin.h [hunk 8]
```diff
@@ -559,18 +579,48 @@ class seq_util {
         app* mk_epsilon(sort* seq_sort);
         info get_info(expr* r) const;
         std::string to_str(expr* r) const;
+        std::string to_strh(expr* r) const;
+
+        expr_ref mk_ite_simplify(expr* c, expr* t, expr* e)
```
Comment (NikolajBjorner) [inline src/ast/seq_decl_plugin.h:584]: use bool_rewriter

File: src/ast/seq_decl_plugin.cpp [hunk 9]
```diff
@@ -1151,100 +1223,209 @@ std::ostream& seq_util::rex::pp::seq_unit(std::ostream& out, expr* s) const {
...
     }
+    else if (re.u.str.is_nth_i(s, e, i)) {
+        print(out, e);
+        out << "[" << mk_pp(i, re.m) << "]";
+    }
+    else if (re.m.is_value(e))
+        out << mk_pp(e, re.m);
+    else if (is_app(e)) {
+        out << "(" << to_app(e)->get_decl()->get_name().str();
+        for (unsigned i = 0; i < to_app(e)->get_num_args(); i++) {
```
[the remark this answers is not in the document]
Comment (veanes) [reply src/ast/seq_decl_plugin.cpp:1234]: fixed

File: src/ast/seq_decl_plugin.h (hunk 8 above)
Comment (veanes) [reply src/ast/seq_decl_plugin.h:584]: The issue here is that `bool_rewriter` may get rid of ite structure, while where `mk_ite_simplify` is being used it is important to keep the if-then-else structure in the **transition-regexes**. So here, I'm keeping it for now.

File: src/ast/rewriter/seq_rewriter.cpp [hunk 10]
```diff
@@ -2976,29 +2992,392 @@ bool seq_rewriter::check_deriv_normal_form(expr* r, int level) {
...
     STRACE("seq_verbose", tout << "derivative result: "
-                               << mk_pp(result, m()) << std::endl;);
+                               << re().to_str(result) << std::endl;);
     CASSERT("seq_regex", check_deriv_normal_form(r));
+    return result;*/
+    return mk_antimirov_deriv(ele, r, m().mk_true());
+}
+
+expr_ref seq_rewriter::mk_antimirov_deriv(expr* e, expr* r, expr* path) {
+    // Take reference count of path and r
```
Comment (NikolajBjorner) [inline src/ast/rewriter/seq_rewriter.cpp:3019]: why not reference count on e?

File: src/ast/rewriter/seq_rewriter.cpp [hunk 11]
```diff
@@ -2976,29 +2992,392 @@ bool seq_rewriter::check_deriv_normal_form(expr* r, int level) {
...
+    else if (re().is_opt(r, r1))
+        result = mk_antimirov_deriv(e, r1, path);
+    else if (re().is_complement(r, r1))
+        // D(e,~r1) = ~D(e,r1)
+        result = mk_antimirov_deriv_negate(mk_antimirov_deriv(e, r1, path));
+    else if (re().is_diff(r, r1, r2))
+        result = mk_antimirov_deriv_intersection(
+            mk_antimirov_deriv(e, r1, path),
+            mk_antimirov_deriv_negate(mk_antimirov_deriv(e, r2, path)), m().mk_true());
+    else if (re().is_of_pred(r, r1)) {
```
Comment (NikolajBjorner) [inline src/ast/rewriter/seq_rewriter.cpp:3165]: hoist to self-contained function that gets used also the other place

File: src/ast/rewriter/seq_rewriter.cpp [hunk 12]
```diff
@@ -2976,29 +2992,392 @@ bool seq_rewriter::check_deriv_normal_form(expr* r, int level) {
...
+        result = (d1->get_id() <= d2->get_id() ? re().mk_inter(d1, d2) : re().mk_inter(d2, d1));
+    return result;
+}
+
+expr_ref seq_rewriter::mk_antimirov_deriv_concat(expr* d, expr* r) {
+    expr_ref result(m());
+    // Take reference count of r and d
+    expr_ref _r(r, m()), _d(d, m());
+    expr* c, * t, * e;
+    if (m().is_ite(d, c, t, e))
```
Comment (NikolajBjorner) [inline src/ast/rewriter/seq_rewriter.cpp:3219]: what is cost model for this rewrite?

**Pull request merged**